File: strivers_a2z/stackandqueue/20validParenthesis.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        if(s.size()&1) return false;
        stack<char> st;
        map<char,char> mp = {{')', '('}, {'}', '{'}, {']', '['}};
        for(char ch:s){
            if(ch=='(' || ch=='{' || ch=='[') st.push(ch);
            else{
                if(st.empty() || mp[ch]!=st.top()) return false;
                else st.pop();
            }
        }
        if(!st.empty()) return false;
        return true;
    }
};

File: NeetCode150/20validParenthesis.py
# https://leetcode.com/problems/valid-parentheses/

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        hashmap = {
            '(': ')',
            '{': '}',
            '[': ']'
        }

        for c in s:
            if c in hashmap:
                stack.append(c)
            else:
                if not stack or c != hashmap[stack.pop()]:
                    return False
        return not stack